### PR TITLE
sync-diff:  make sure user setting can cover the default same table route rules

### LINKED ISSUE #558: sync-diff-inspector: mysql to mysql check error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

Create two databases and tables in one MySQL instance.

```
CREATE DATABASE inspector_mysql_0;
CREATE TABLE inspector_mysql_0.tb_emp1 (id INT(11), name VARCHAR(25), deptId INT(11));
INSERT INTO inspector_mysql_0.tb_emp1 values (1,"hello1",1);
CREATE DATABASE inspector_mysql_1;
CREATE TABLE inspector_mysql_1.tb_emp1 (id INT(11), name VARCHAR(25), deptId INT(11));
INSERT INTO inspector_mysql_1.tb_emp1 values (1,"hello1",1);
```

then run sync-diff-inspector with config.toml

```toml
# Diff Configuration.

######################### Global config #########################

# how many goroutines are created to check data
check-thread-count = 4

# set false if just want compare data by checksum, will skip select data when checksum is not equal.
# set true if want compare all different rows, will slow down the total compare time.
export-fix-sql = true

# ignore check table's data
check-struct-only = false


######################### Databases config #########################
[data-sources.mysql1]
    host = "127.0.0.1"
    port = 3306
    user = "root"
    password = "12344321"
    route-rules = ["rule1"]

[data-sources.tidb0]
    host = "127.0.0.1"
    port = 3306
    user = "root"
    password = "12344321"
    # remove comment if use tidb's snapshot data

[routes.rule1]
schema-pattern = "inspector_mysql_0"      # schema to match. Support wildcard characters * and ?.
table-pattern = "tb_emp1"          # table to match. Support wildcard characters * and ?.
target-schema = "inspector_mysql_1"         # target schema
target-table = "tb_emp1"             # target table

######################### Task config #########################
# Required
[task]
    # 1 fix sql: fix-target-TIDB1.sql
    # 2 log: sync-diff.log
    # 3 summary: summary.txt
    # 4 checkpoint: a dir
    output-dir = "/tmp/output/config"

    source-instances = ["mysql1"]

    target-instance = "tidb0"

    # tables need to check. *Include `schema` and `table`. Use `.` to split*
    target-check-tables = ["inspector_mysql_1.tb_emp1"]

    # extra table config
#     target-configs= []
    target-configs= ["config1"]

# Optional
# [table-configs]
[table-configs.config1]

# tables need to use this specified config.
# if use this config. target-tables should be a subset of #target-check-tables
target-tables = ["inspector_mysql_1.tb_emp1"]

range = "id > 0"
index-fields = [""]
ignore-columns = ["",""]
chunk-size = 0
collation = ""
```

2. What did you expect to see?

```
The table structure and data in following tables are equivalent

`inspector_mysql_1`.`tb_emp1`
```

3. What did you see instead?

```
The table structure and data in following tables are equivalent


The following tables contains inconsistent data

+-------------------------------+--------------------+----------------+
|             TABLE             | STRUCTURE EQUALITY | DATA DIFF ROWS |
+-------------------------------+--------------------+----------------+
| `inspector_mysql_1`.`tb_emp1` | true               | +1/-0          |
+-------------------------------+--------------------+----------------+
```

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?

MySQL 8.0.22

5. which tool are you using?

tidb-tools/sync-diff-inspector

6. what version of tool are you using (`pump -V` or `tidb-lightning -V` or `syncer -V`)?

```
App Name: sync_diff_inspector v2.0
Release Version: v5.2.2-113-g39e3eb4-dirty
Git Commit Hash: 39e3eb4bb9725bf714bd6c23a77de69852ea2a27
Git Branch: master
UTC Build Time: 2021-11-28 07:25:14
Go Version: go1.17
```

### PR DESCRIPTION
**Pull request by 3pointer**
### What problem does this PR solve?
Another way to solve #558 

Issue Number: close #558

### What is changed and how it works?
Make sure user setting can cover the default rule setting.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

image

 - No code

Code changes

 - Has exported function/method change
 - Has exported variable/fields change
 - Has interface methods change
 - Has persistent data change

Side effects

 - Possible performance regression
 - Increased code complexity
 - Breaking backward compatibility

Related changes

 - Need to cherry-pick to the release branch
 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Review by lichunzhu — APPROVED** [review]
LGTM

**Review by lichunzhu — APPROVED** [review]
LGTM

**Review by lance6716 — APPROVED** [review]

**Comment by Leavrth** [comment]
PTAL https://github.com/pingcap/tidb-tools/pull/612

File: sync_diff_inspector/source/source.go [hunk 1]
```diff
@@ -134,13 +139,48 @@ func NewSources(ctx context.Context, cfg *config.Config) (downstream Source, ups
...
 		for _, d := range cfg.Task.SourceInstances {
-			if d.Router.AddRule(&router.TableRule{
-				SchemaPattern: tableConfig.Schema,
-				TablePattern:  tableConfig.Table,
-				TargetSchema:  tableConfig.Schema,
-				TargetTable:   tableConfig.Table,
-			}) != nil {
-				return nil, nil, errors.Errorf("set case unsensitive failed. The schema/table name cannot be parttern. [schema = %s] [table = %s]", tableConfig.Schema, tableConfig.Table)
+			if _, ok := d.RouteTargetSet[dbutil.TableName(tableConfig.Schema, tableConfig.Table)]; ok {
+				// There is a user rule routing to `tableConfig.Schema`.`tableConfig.Table`
```
Comment (lichunzhu) [inline sync_diff_inspector/source/source.go:143]: Maybe we can add a case in comment here to describe the behavior here:

| sourceDB | targetDB |
| :---: | :---: |
| test1 | test |
| test2 | test2 |         
| test | not int routes |

In this case we will check test1 -> test, test2 -> test2. But we won't check schema test.

File: sync_diff_inspector/source/source_test.go [hunk 2]
```diff
@@ -685,6 +685,106 @@ func TestSource(t *testing.T) {
...
+	cfg.Task.TargetCheckTables, err = filter.Parse([]string{"schema2.tbl"})
+	require.NoError(t, err)
+	cfg.Task.SourceInstances[0].RouteTargetSet[dbutil.TableName("schema2", "tbl")] = struct{}{}
+
+	// create table
+	conn, err := sql.Open("mysql", fmt.Sprintf("root:@tcp(%s:%d)/?charset=utf8mb4", host, port))
+	require.NoError(t, err)
+
+	conn.Exec("CREATE DATABASE IF NOT EXISTS schema1")
+	conn.Exec("CREATE TABLE IF NOT EXISTS `schema1`.`tbl` (`a` int, `b` varchar(24), `c` float, `d` datetime, primary key(`a`, `b`))")
```
Comment (lichunzhu) [inline sync_diff_inspector/source/source_test.go:767]: Do we need to drop this table/schema first?

File: sync_diff_inspector/source/source_test.go [hunk 3]
```diff
@@ -685,6 +685,106 @@ func TestSource(t *testing.T) {
...
+	r, err := router.NewTableRouter(false, []*router.TableRule{
+		// make sure this rule works
+		{
+			SchemaPattern: "schema1",
+			TablePattern:  "tbl",
+			TargetSchema:  "schema2",
+			TargetTable:   "tbl",
+		},
+	})
+	cfg := &config.Config{
```
Comment (lichunzhu) [inline sync_diff_inspector/source/source_test.go:712]: What about using a toml configuration string here?

File: tests/sync_diff_inspector/shard/run.sh [hunk 4]
```diff
@@ -35,4 +35,100 @@ sync_diff_inspector --config=./config.toml > $OUT_DIR/shard_diff.output || true
...
+# router_test_1.Tbl
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "create database if not exists router_test_1;"
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "create table router_test_1.Tbl (id INT(11), name VARCHAR(25), deptId INT(11));"
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "insert into router_test_1.Tbl values (1,\"hello1\",1);"
+# Router_test_1.Tbl
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "create database if not exists Router_test_1;"
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "create table Router_test_1.Tbl (id INT(11), name VARCHAR(25), deptId INT(11));"
+mysql -uroot -h ${MYSQL_HOST} -P ${MYSQL_PORT} -e "insert into Router_test_1.Tbl values (1,\"hello1\",1);"
+
+echo "test router 1: normal rule"
```
Comment (lichunzhu) [inline tests/sync_diff_inspector/shard/run.sh:68]: Can we use a function to simplify this part?

File: sync_diff_inspector/source/source_test.go (hunk 2 above)
Comment (3pointer) [reply sync_diff_inspector/source/source_test.go:767]: we need keep this `schema1`.`tbl` for next router check.

File: sync_diff_inspector/source/source_test.go (hunk 3 above)
Comment (3pointer) [reply sync_diff_inspector/source/source_test.go:712]: sure, but we can fix it in another PR. cause other test case has the same code.

File: sync_diff_inspector/source/source.go (hunk 1 above)
Comment (3pointer) [reply sync_diff_inspector/source/source.go:143]: sure, add a unit test case in https://github.com/pingcap/tidb-tools/pull/611/commits/267c3d6bc7e79ca41d1936bb1793b58907bea109. PTAL

**Review by lichunzhu — APPROVED** [review]

**Pull request merged**